### PR TITLE
Add separate cuda and nvrtc initialization, refactor generator code

### PR DESCRIPTION
**Pull request by brechtvl**
These new functions can be called to use nvrtc without CUDA hardware available, since `cuewInit()` can fail if that's not the case.

Changing the generator code to add these was a pain though. Instead I've made that code use cuew.template.c/h files. These are easier to edit, inserting autogenerated code in a few places rather than generating all C code through Python.

### CODE REVIEW

**Comment by brechtvl** [comment]
I forgot to mention: this also updates the nvrtc dll names, and part of this patch was written by @LazyDodo.

**Comment by sergeyvfx** [comment]
Part related on the template is something which will become handy in the future.

But part of split `cuewInit()` and `cuewNvrtcInit()` i didn't quite understand. We don't require hardware, and only requirement i see is CUDA driver version to be 4.0 at least. I don't think older drivers had NVRCT in them, and how is it possible that you have NVRTC without newer driver installed? Has it became part of display driver instead?

To me it seems we should have single `cuewInit()` function which will initialize all available wranglers. If it is REALLY needed to specify which parts we want to ignore, make it a bitmask argument to it.

Put all this information into a commit message. Having such a brief oneliners do not help at all. Also, don't use fullstop in commit subject (first line of commit message).

**Review by sergeyvfx — CHANGES_REQUESTED** [review]
More information needed, see the comment in conversation.

**Comment by brechtvl** [comment]
`libcuda` is part of the driver, and `dlopen()` on it fails there is no CUDA hardware. And `libnvrtc` can be installed as part of the CUDA toolkit, without the driver. What we are interested in is loading `libnvrtc` to compile programs without having any CUDA hardware.

If we add a argument to `cuewInit()`, we break backwards compatibility. I wouldn't mind doing that though, because I think it's bad that we are now automatically loading `libnvrtc` along with `libcuda`. Most of the time there is no reason to have the time and memory overhead of loading the compiler when all you need to do is render something.

**Comment by sergeyvfx** [comment]
> What we are interested is loading libnvrtc to compile programs without having any CUDA hardware.

What is the usecase for that?

> If we add a argument to cuewInit(), we break backwards compatibility.

You break it anyway by removing NVRTC library load from `cuewInit()`. So everyone will need to change their programs anyway. I also not sure we ever declared API stable. We didn't even make a release or anything..

If you are really worried about compatibility, it's possible to keep compatibility by doing some variadic macro, which will convert `cuewInit()` to a call with both CUDA and NVRTC requested for load. But i don't think it worth it.

> Most of the time there is no reason to have the time and memory overhead of loading the compiler when all you need to do is render something.

Did you quantify memory and time? Is it really that bad?

**Comment by brechtvl** [comment]
> What is the usecase for that?

Compiling CUDA binaries for Blender releases: https://developer.blender.org/D3045

> Did you quantify memory and time? Is it really that bad?

No, I did not quantify it, and it may not be bad in practice. But it's a big library, it might be doing a bunch of static initialization, and these things can change over time. It's just an argument in favor or breaking backwards compatibility, which you already seem to agree with so it doesn't matter.

**Comment by sergeyvfx** [comment]
> Compiling CUDA binaries for Blender releases: https://developer.blender.org/D3045

Makes sense. But all such usecases are to be reflected in clear explanatory commit message.

> No, I did not quantify it, and it may not be bad in practice. But it's a big library, it might be doing a bunch of static initialization, and these things can change over time. It's just an argument in favor or breaking backwards compatibility, which you already seem to agree with so it doesn't matter.

Yes, make it a bitmask argument to cuewInit(). Internally you can still split the function up, to avoid too long of a function which does everything.

**Comment by brechtvl** [comment]
Updated with flags for `cuewInit()`, version bump to indicate compatibility breakage, and better commit messages.

File: auto/cuew.template.c [hunk 1]
```diff
@@ -156,15 +145,87 @@ int cuewInit(void) {
...
+static void cuewExitNvrtc(void) {
   if (nvrtc_lib != NULL) {
-%LIB_FIND_NVRTC%
+    /*  Ignore errors. */
+    dynamic_library_close(nvrtc_lib);
+    nvrtc_lib = NULL;
+  }
+}
+
+int cuewNvrtcInit(void) {
```
Comment (sergeyvfx) [inline auto/cuew.template.c]: Seems to be private function which is not exposed to a public API. Hence should be using static qualifier.

**Review by sergeyvfx — CHANGES_REQUESTED** [review]

**Comment by brechtvl** [comment]
Made function static.

**Review by sergeyvfx — APPROVED** [review]
Think we are good now.

Do you have commit access here?

**Comment by brechtvl** [comment]
No, I wouldn't mind getting it though.

**Pull request merged**